### PR TITLE
toffset to timestamp

### PR DESCRIPTION
**Pull request by guperrot**

### CODE REVIEW

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/channel/DefaultChannel.java [hunk 1]
```diff
@@ -527,8 +528,8 @@ public synchronized void enqueue(@NonNull Log log, @NonNull final String groupNa
         }
 
         /* Set an absolute timestamp, we'll convert to relative just before sending. Don't do it if the service already set a timestamp.*/
```
Comment (dhei) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/channel/DefaultChannel.java]: Should we update this code comment?

Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/channel/DefaultChannel.java]: Oh, yes

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 2]
```diff
@@ -449,7 +448,7 @@ public void onCallBack(ErrorReport report) {
     /**
      * Get initialization timestamp.
      *
-     * @return initialization timestamp expressed using {@link SystemClock#elapsedRealtime()}.
+     * @return initialization timestamp expressed using {@link System#currentTimeMillis()}}.
```
Comment (dhei) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: typo: `}}` should be `}` at the end

**Comment by guperrot** [comment]
Must wait backend fix on `app_launch_timestamp` before merging.

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java [hunk 3]
```diff
@@ -177,12 +106,65 @@ private static void writePropertyValue(JSONStringer writer, Object value) throws
...
+     * @return the properties value
+     */
+    public Map<String, Object> getProperties() {
+        return this.properties;
+    }
+
+    /**
+     * Set the properties value.
+     *
+     * @param properties the properties value to set
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java]: missing dot

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java [hunk 4]
```diff
@@ -177,12 +106,65 @@ private static void writePropertyValue(JSONStringer writer, Object value) throws
...
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Get the properties value.
+     *
+     * @return the properties value
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java]: missing dot

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java [hunk 5]
```diff
@@ -12,18 +13,18 @@
...
      */
-    long getToffset();
+    Date getTimestamp();
 
     /**
-     * Set the toffset value.
+     * Set the timestamp value.
      *
-     * @param toffset the toffset value to set
+     * @param timestamp the timestamp value to set
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java]: missing dot

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java [hunk 6]
```diff
@@ -12,18 +13,18 @@
     String getType();
 
     /**
-     * Get the toffset value.
+     * Get the timestamp value.
      *
-     * @return the toffset value
+     * @return the timestamp value
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java]: missing dot

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java [hunk 7]
```diff
@@ -230,21 +233,21 @@ public void setFatal(Boolean fatal) {
...
+    public Date getAppLaunchTimestamp() {
+        return this.appLaunchTimestamp;
     }
 
     /**
-     * Set the appLaunchTOffset value.
+     * Set the appLaunchTimestamp value.
      *
-     * @param appLaunchTOffset the appLaunchTOffset value to set
+     * @param appLaunchTimestamp the appLaunchTimestamp value to set
```
Comment (MatkovIvan) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: missing dot

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java [hunk 8]
```diff
@@ -230,21 +233,21 @@ public void setFatal(Boolean fatal) {
     }
 
     /**
-     * Get the appLaunchTOffset value.
+     * Get the appLaunchTimestamp value.
      *
-     * @return the appLaunchTOffset value
+     * @return the appLaunchTimestamp value
```
Comment (MatkovIvan) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: missing dot

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java [hunk 9]
```diff
@@ -19,24 +21,27 @@
 public abstract class AbstractLog implements Log {
 
     /**
+     * timestamp property.
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java]: Capital letter or quotes

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java [hunk 10]
```diff
@@ -19,24 +21,27 @@
...
+    static final String TIMESTAMP = "timestamp";
+
+    /**
      * Session identifier property.
      */
     @VisibleForTesting
     static final String SID = "sid";
+
     /**
      * device property.
```
Comment (MatkovIvan) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java]: Capital letter or quotes

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java [hunk 11]
```diff
@@ -171,6 +173,7 @@ public String getParentProcessName() {
      *
      * @param parentProcessName the parentProcessName value to set
```
Comment (MatkovIvan) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: missing dot

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java [hunk 12]
```diff
@@ -153,6 +154,7 @@ public Integer getParentProcessId() {
      *
      * @param parentProcessId the parentProcessId value to set
```
Comment (MatkovIvan) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: missing dot

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java (hunk 4 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java]: We started off the generated code from autorest for Java then edited those files, hence they have a different style (the json ones) from other files.

Will refactor every JSON file at once someday but this PR is out of scope.

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java (hunk 3 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/CustomPropertiesLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java (hunk 6 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java (hunk 5 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/Log.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java (hunk 12 above)
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java (hunk 11 above)
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java (hunk 8 above)
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java (hunk 7 above)
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/ingestion/models/AbstractErrorLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java (hunk 9 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

File: sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java (hunk 10 above)
Comment (guperrot) [inline sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/ingestion/models/AbstractLog.java]: https://github.com/Microsoft/mobile-center-sdk-android/pull/481#discussion_r127363145

**Comment by jaelim-ms** [comment]
Have you considered to set the time in constructor in addition to the setter? I think you can either pass a Date instance as a parameter or just set it as by default.

**Comment by jaelim-ms** [comment]
Workaround is ready on PROD.
